include/swift/Reflection/ReflectionContext.h

```diff
@@ -112,83 +103,119 @@ class ReflectionContext
-    auto TyperefMd = findSection<GenericSection>(Header, "__swift5_typeref");
-    auto ReflStrMd = findSection<GenericSection>(Header, "__swift5_reflstr");
-
-    bool success = FieldMd.second || AssocTyMd.second || BuiltinTyMd.second ||
-                   CaptureMd.second || TyperefMd.second || ReflStrMd.second;
-    if (!success)
+    auto findMachOSectionByName = [&](std::string Name)
+        -> std::pair<std::pair<const char *, const char *>, uint32_t> {
+      // Now for all the sections, find their name.
+      for (auto &S : SectionsHdr) {
+        if (strncmp(S->sectname, Name.c_str(), strlen(Name.c_str())) != 0)
+          continue;
+        auto SecStart = RemoteAddress(ImageStart.getAddressData() + S->offset);
+        auto SecSize = S->size;
+        auto SecBuf = this->getReader().readBytes(SecStart, SecSize);
+        auto SecContents = reinterpret_cast<const char *>(SecBuf.get());
+        return {{SecContents, SecContents + SecSize}, 0};
+      }
+      return {{nullptr, nullptr}, 0};
+    };
+
+    auto FieldMdSec = findMachOSectionByName("__swift5_fieldmd");
+    auto AssocTySec = findMachOSectionByName("__swift5_assocty");
+    auto BuiltinTySec = findMachOSectionByName("__swift5_builtin");
+    auto CaptureSec = findMachOSectionByName("__swift5_capture");
+    auto TypeRefMdSec = findMachOSectionByName("__swift5_typeref");
+    auto ReflStrMdSec = findMachOSectionByName("__swift5_reflstr");
+
+    if (FieldMdSec.first.first == nullptr &&
+        AssocTySec.first.first == nullptr &&
+        BuiltinTySec.first.first == nullptr &&
+        CaptureSec.first.first == nullptr &&
+        TypeRefMdSec.first.first == nullptr &&
+        ReflStrMdSec.first.first == nullptr)
       return false;
 
+    Buf = this->getReader().readBytes(ImageStart, sizeof(typename T::Header));
     auto LocalStartAddress = reinterpret_cast<uintptr_t>(Buf.get());
     auto RemoteStartAddress = static_cast<uintptr_t>(ImageStart.getAddressData());
 
     ReflectionInfo info = {
-        {{FieldMd.first.startAddress(), FieldMd.first.endAddress()}, 0},
-        {{AssocTyMd.first.startAddress(), AssocTyMd.first.endAddress()}, 0},
-        {{BuiltinTyMd.first.startAddress(), BuiltinTyMd.first.endAddress()}, 0},
-        {{CaptureMd.first.startAddress(), CaptureMd.first.endAddress()}, 0},
-        {{TyperefMd.first.startAddress(), TyperefMd.first.endAddress()}, 0},
-        {{ReflStrMd.first.startAddress(), ReflStrMd.first.endAddress()}, 0},
+        {{FieldMdSec.first.first, FieldMdSec.first.second}, 0},
+        {{AssocTySec.first.first, AssocTySec.first.second}, 0},
+        {{BuiltinTySec.first.first, BuiltinTySec.first.second}, 0},
+        {{CaptureSec.first.first, CaptureSec.first.second}, 0},
+        {{TypeRefMdSec.first.first, TypeRefMdSec.first.second}, 0},
+        {{ReflStrMdSec.first.first, ReflStrMdSec.first.second}, 0},
         LocalStartAddress,
         RemoteStartAddress};
```

Comment:
I'm not quite sure that `LocalStartAddress` and `RemoteStartAddress` make sense anymore. My (limited) understanding is that they are used to translate between an address in the local space and an address in the remote space, which assumes that all of the reflection sections have the same delta between the two address spaces. That was true when the entire TEXT segment was being mapped in one shot, but is not necessarily true with these changes.

It looks like the only code that actually uses those addresses is `getRemoteAddrOfTypeRefPointer` in `TypeRefBuilder.cpp`. It only translates addresses within the `TypeReference` sections. Maybe we can get what we need by just storing the remote address of that one section.

Reply:
I spoke with Nick and he recommended to just map the whole `__TEXT` as taking in account the delta between sections may be complicated. This is what I did. If this turns out to be a bottleneck we might reconsider it but I really want to make some progress on the debugger integration.

Reply:
Makes sense to me. Is it actually slower to map everything in one chunk compared to mapping several smaller chunks? I'd have expected each mapping operation to be approximately constant time regardless of size, but I could certainly be wrong.